core/interop/src/lib.rs

```diff
@@ -40,67 +46,251 @@ impl<T: IntoIterator<Item = (JsString, NativeFunction)> + Clone> IntoJsModule fo
-        unsafe {
-            NativeFunction::from_closure(move |_, _, _| {
-                cell.borrow_mut()();
-                Ok(JsValue::undefined())
-            })
+/// The safe equivalent of the [`UnsafeIntoJsFunction`] trait.
+/// This can only be used on closures that have the `Copy` trait.
+///
+/// Since this function is implemented for `Fn(...)` closures, we can use
+/// it directly when defining a function:
+/// ```
+/// # use boa_engine::{Context, JsValue, NativeFunction};
+/// # use boa_interop::IntoJsFunctionCopied;
+/// # let mut context = Context::default();
+/// let f = |a: i32, b: i32| a + b;
+/// let f = f.into_js_function(&mut context);
+/// let result = f.call(&JsValue::undefined(), &[JsValue::from(1), JsValue::from(2)], &mut context).unwrap();
+/// assert_eq!(result, JsValue::new(3));
+/// ```
+pub trait IntoJsFunctionCopied<Args, Ret>: private::IntoJsFunctionSealed<Args, Ret> + Copy {
+    /// Converts the type into a JS function.
+    fn into_js_function(self, context: &mut Context) -> NativeFunction;
+}
+
+/// Create a [`JsResult`] from a Rust value. This trait is used to
+/// convert Rust types to JS types, including [`JsResult`] of
+/// Rust values and [`JsValue`]s. It is used to convert the
+/// return value of a function in [`UnsafeIntoJsFunction`] and
+/// [`IntoJsFunctionCopied`].
+pub trait TryIntoJsResult {
+    /// Try to convert a Rust value into a `JsResult<JsValue>`.
+    ///
+    /// # Errors
+    /// Any parsing errors that may occur during the conversion, or any
+    /// error that happened during the call to a function.
+    fn try_into_js_result(self, context: &mut Context) -> JsResult<JsValue>;
+}
+
+mod try_into_js_result_impls;
+
+/// Create a Rust value from a JS argument. This trait is used to
+/// convert arguments from JS to Rust types. It allows support
+/// for optional arguments or rest arguments.
+pub trait TryFromJsArgument: Sized {
+    /// Try to convert a JS argument into a Rust value, returning the
+    /// value and the rest of the arguments to be parsed.
+    ///
+    /// # Errors
+    /// Any parsing errors that may occur during the conversion.
+    fn try_from_js_argument<'a>(
+        this: &'a JsValue,
+        rest: &'a [JsValue],
+        context: &mut Context,
+    ) -> JsResult<(Self, &'a [JsValue])>;
+}
+
+impl<T: TryFromJs> TryFromJsArgument for T {
+    fn try_from_js_argument<'a>(
+        _: &'a JsValue,
+        rest: &'a [JsValue],
+        context: &mut Context,
+    ) -> JsResult<(Self, &'a [JsValue])> {
+        match rest.split_first() {
+            Some((first, rest)) => Ok((first.try_js_into(context)?, rest)),
+            None => T::try_from_js(&JsValue::undefined(), context).map(|v| (v, rest)),
         }
     }
 }
 
+/// An argument that when used in a JS function will empty the list
+/// of JS arguments as `JsValue`s. This can be used for having the
+/// rest of the arguments in a function.
+#[derive(Debug, Clone)]
+pub struct JsRest(pub Vec<JsValue>);
```

Comment:
Does this need to be a `Vec<JsValue>`? I would argue it would be better to have it be `&[JsValue]` to avoid unnecessary allocations, but I don't know if there are technical challenges that disallow this.

Reply:
It turns out that it becomes a bit convoluted for the trait implementation. I'd need to make changes to `NativeFunction`.

Reply:
We can do incremental improvements later then.